### ISSUE TITLE
Support Client Certificates Without SAN

### ISSUE THREAD

**Issue by henryreed** [caddyserver/caddy#5558]
Hi folks,

Client certificates without a SAN are not allowed to authenticate to Caddy; Caddy instead returns back an error, `{"level":"debug","ts":1685268475.3135417,"logger":"http.stdlib","msg":"http: TLS handshake error from 192.168.3.2:40411: tls: invalid signature by the client certificate: ECDSA verification failure"}`

I would prefer Caddy not return that error and instead let SAN-less certificates authenticate because I would like to authenticate with the PIV application on my Yubikey. It is not possible to generate a CSR with a SAN on a Yubikey. While I can generate my own certificate with a SAN and load it onto the Yubikey, I would prefer the private key to be generated on the Yubikey itself, which is why I'd like the SAN-less client certificate authentication support to be added to Caddy.

A Caddyfile as an example (note that in my own environment, I am using a third-party CA for the HTTPS connection on a public DNS record):
```
{
    log {
        level debug
    }
}

(client_auth) {
    tls {
        client_auth {
            mode require_and_verify
            trusted_ca_cert_file /certs/root-ca.crt
            trusted_ca_cert_file /certs/intermediate-ca.crt
            trusted_leaf_cert_file /certs/leaf.crt
            trusted_leaf_cert_file /certs/sad_and_invalid_leaf.crt
        }
    }
}

test.local {
    import client_auth
    respond "hello darkness my old friend"
}
```

How I generate a valid, SAN-ful client certificate with Smallstep CLI that can successfully authenticate with the above Caddyfile:

```bash
step certificate create "Please Help Me Root CA" root-ca.crt root-ca.key --profile root-ca --not-after 262980h --crv=P-384 --kty=EC
step certificate create "Please Help Me Intermediate CA" intermediate-ca.crt intermediate-ca.key --profile intermediate-ca --ca ./root-ca.crt --ca-key ./root-ca.key --not-after 87660h --crv=P-384 --kty=EC
step certificate create "Please Help Me Leaf Certificate" leaf.crt leaf.key --profile leaf --ca intermediate-ca.crt --ca-key intermediate-ca.key --not-after 17532h --bundle --crv=P-384 --kty=EC
```

Note the existence of a SAN in this certificate:
```bash
$ step certificate inspect leaf.crt | grep -A1 Alternative
            X509v3 Subject Alternative Name:
                DNS:Please Help Me Leaf Certificate
```

How I sign a CSR from a Yubikey, which is ultimately rejected by Caddy:

```bash
step certificate sign --bundle --not-after 17532h leaf.csr intermediate-ca.crt intermediate-ca.key >> sad_and_invalid_leaf.crt
```

Note the nonexistence of a SAN in this certificate:
```bash
$ step certificate inspect sad_and_invalid_leaf.crt | grep -A1 Alternative

```

Please let me know if you need me to provide more information or if I missed something. I believe the existence/nonexistence of a SAN is ultimately the reason why Yubikey-CSR-derived leaf certificates are rejected, but I am open to listening to alternative possibilities, also.

**Comment by francislavoie** [comment]
That error is coming from Go stdlib, not from Caddy itself:

https://github.com/golang/go/blob/7ad92e95b56019083824492fbec5bb07926d8ebd/src/crypto/tls/handshake_server.go#L723

https://github.com/golang/go/blob/7ad92e95b56019083824492fbec5bb07926d8ebd/src/crypto/tls/auth.go#L30

**Comment by henryreed** [comment]
Thanks for looking into this @francislavoie. Unfortunately, I am not quite sure what is going on, and why it "didn't work before but it does now" (classic, I know). I regenerated a CSR on my Yubikey, signed it again, added it to the cert store in Caddy and it works fine without errors. I am unable to repeat the issue I described here. I will close the issue, but if anyone else runs into this: try regenerating a new CSR, signing it, and adding it to Caddy; maybe also pray to the crypto gods.
